arch/xtensa/core/gen_vectors.py

```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/env python3
+# Copyright 2023 The ChromiumOS Authors
+# SPDX-License-Identifier: Apache-2.0
+import sys
+import re
+
+# Xtensa Vector Table linker generator
+#
+# Takes a pre-processed (gcc -dM) core-isa.h file as its first
+# argument, and emits a GNU linker section declartion which will
+# correctly load the exception vectors and literals as long as their
+# code is declared using standard conventions (see below).
+#
+# The section name will be ".z_xtensa_vectors", and a symbol
+# "z_xtensa_vecbase" is emitted containing a valid value for the
+# VECBASE SR at runtime.
+#
+# Obviously, this requires that XCHAL_HAVE_VECBASE=1.  A similar trick
+# could be played to load vectors at fixed addresses on hardware that
+# lacks VECBASE, but the core-isa.h interface is inexplicably
+# different.
+#
+# Because the "standard conventions" (which descend from somewhere in
+# Cadence) are not documented anywhere and just end up cut and pasted
+# between devices, here's an attempt at a specification:
+#
+# + The six register window exception vectors are defined with offsets
+#   internal to their assembly code.  They are linked in a single
+#   section named ".WindowVectors.text".
+#
+# + The "kernel", "user" and "double exception" vectors are emitted in
+#   sections named ".KernelExceptionVector.text",
+#   "UserExceptionVector.text" and "DoubleExceptionVector.text"
+#   respectively.
+#
+# + XEA2 interrupt vectors are in sections named
+#   ".Level<n>InterruptVector.text", except (!) for ones which are
+#   given special names.  The "debug" and "NMI" interrupts (if they
+#   exist) are technically implemented as standard interrupt vectors
+#   (of a platform-dependent level), but the code for them is emitted
+#   in ".DebugExceptionVector.text" and ".NMIExceptionVector.text",
+#   and not a section corresponding to their interrupt level.
+#
+# + Any unused bytes at the end of a vector are made available as
+#   storage for immediate values used by the following vector (Xtensa
+#   can only back-reference immediates for MOVI/L32R instructions) as
+#   a "<name>Vector.literal" section.  Note that there is no guarantee
+#   of how much space is available, it depends on the previous
+#   vector's code size.  Zephyr code has historically not used this
+#   space, as support in existing linker scripts is inconsistent.  But
+#   it's exposed here.
+
+coreisa = sys.argv[1]
+
+debug_level = 0
```

Comment:
Yeah I think this script could use some Pythonification but I imagine it may be a followup, worth fixing it up with `gen_vectors.py` and `gen_zsr.py` as well